{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr1\n",
      "Chr2\n",
      "Chr3\n",
      "Chr4\n",
      "Chr5\n",
      "Chr6\n",
      "Chr7\n",
      "Chr8\n",
      "Chr9\n",
      "Chr10\n",
      "Chr11\n",
      "Chr12\n",
      "ChrUn\n",
      "ChrSy\n"
     ]
    }
   ],
   "source": [
    "# transcript_location_to_genome_location()\n",
    "gff=\"/home/wuyou/Projects/Osativa_nanopore/ref/Osativa_323_v7.0.gene_exons.gtf\"\n",
    "\n",
    "fasta=\"/home/wuyou/Projects/Osativa_nanopore/ref/Osativa_323_v7.0.fa\"\n",
    "\n",
    "sequence_dict={}\n",
    "\n",
    "\n",
    "chr=\"Chr0\"\n",
    "sequence=\"\"\n",
    "with open(fasta) as f:\n",
    "    for line in f:\n",
    "\n",
    "        line=line.rstrip()\n",
    "        if \">\" in line:\n",
    "            sequence_dict[chr]=sequence\n",
    "            sequence=\"\"\n",
    "            \n",
    "            chr=line[1:]\n",
    "            print(chr) \n",
    "        else:\n",
    "            sequence+=line\n",
    "sequence_dict[chr]=sequence    \n",
    "\n",
    "\n",
    "\n",
    "exon_dict={}\n",
    "chr_dict={}\n",
    "strand_dict={}\n",
    "with open(gff) as f:\n",
    "    for line in f:\n",
    "        type=line.split(\"\\t\")[2]\n",
    "        if type==\"exon\":\n",
    "            chr=line.split('\\t')[0]\n",
    "            transcript_id=line.split('\"')[1]\n",
    "            start=int(line.split(\"\\t\")[3])\n",
    "            end=int(line.split(\"\\t\")[4])\n",
    "            strand=line.split(\"\\t\")[6]\n",
    "            if transcript_id not in exon_dict:\n",
    "                exon_dict[transcript_id]=[]\n",
    "                chr_dict[transcript_id]=chr\n",
    "                strand_dict[transcript_id]=strand\n",
    "            exon_dict[transcript_id].append([start,end])\n",
    "\n",
    "\n",
    "def revere_complement(seq):\n",
    "    seq=seq.replace(\"A\",\"t\")\n",
    "    seq=seq.replace(\"C\",\"g\")\n",
    "    seq=seq.replace(\"G\",\"c\")\n",
    "    seq=seq.replace(\"T\",\"a\")\n",
    "    return seq.upper()[::-1]\n",
    "\n",
    "\n",
    "import time\n",
    "transcript_location=\"../demo_data/test3.prediction.tsv\"\n",
    "genome_location=\"../demo_data/test3.prediction.genome_loc.tsv\"\n",
    "out=open(genome_location,\"w\")\n",
    "\n",
    "count=0\n",
    "with open(transcript_location) as f:\n",
    "    for line in f:\n",
    "        count+=1\n",
    "        if count % 1e7==0:\n",
    "            print(count)\n",
    "        items=line.rstrip().split(\"\\t\")\n",
    "        transcript_id=items[0]\n",
    "        transcript_location=int(items[1])\n",
    "        motif=items[2]\n",
    "        mod=items[4]\n",
    "        probability=items[5]\n",
    "        \n",
    "        exons=exon_dict[transcript_id]\n",
    "        strand=strand_dict[transcript_id]\n",
    "        if strand==\"+\":\n",
    "            summation=0\n",
    "            for exon in exons:\n",
    "                start=exon[0]\n",
    "                end=exon[1]\n",
    "                if transcript_location>summation+end-start+1:\n",
    "                    summation=summation+end-start+1\n",
    "                else:\n",
    "                    chr_location=transcript_location-summation+start-1\n",
    "                    break\n",
    "\n",
    "            print(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\" %(transcript_id,transcript_location,motif,chr_dict[transcript_id],chr_location,sequence_dict[chr_dict[transcript_id]][chr_location-3:chr_location+2],mod,probability),file=out)\n",
    "        if strand==\"-\":\n",
    "            summation=0\n",
    "            for exon in exons[::-1]:       #reverse\n",
    "                start=exon[0]\n",
    "                end=exon[1]\n",
    "                if transcript_location>summation+end-start+1:\n",
    "                    summation=summation+end-start+1\n",
    "                else:\n",
    "                    chr_location=end-(transcript_location-summation)+1\n",
    "                    break\n",
    "\n",
    "            print(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\" %(transcript_id,transcript_location,motif,chr_dict[transcript_id],chr_location,revere_complement(sequence_dict[chr_dict[transcript_id]][chr_location-3:chr_location+2]),mod,probability),file=out)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TandemMod",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
